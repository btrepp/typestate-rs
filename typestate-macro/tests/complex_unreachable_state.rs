use typestate::typestate;

#[typestate]
mod traffic_light {
    #[automata]
    pub struct TrafficLight {
        pub cycles: u64,
    }

    #[state]
    pub struct Green;

    #[state]
    pub struct Yellow;

    #[state]
    pub struct Red;

    #[state]
    pub struct RedA;

    pub trait Green {
        fn to_yellow(self) -> Yellow;
    }

    pub trait Yellow {
        fn to_red(self) -> Red;
    }

    pub trait Red {
        fn to_green(self) -> Green;
        fn turn_on() -> Red;
        fn turn_off(self);
    }
}

fn main() {}
