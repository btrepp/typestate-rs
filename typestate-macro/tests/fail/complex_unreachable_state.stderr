error: Non-productive state. For a state to be productive, a path from the state to a final state is required to exist.
  --> $DIR/complex_unreachable_state.rs:20:16
   |
20 |     pub struct RedA;
   |                ^^^^
